# Lab Assignment 2

## Part 1: Multinomial Naive Bayes with CountVectorizer

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

df = pd.read_csv('spam.csv', encoding='latin-1')
df = df.drop(df.iloc[:, 2:], axis=1)
df.columns = ['Labels', 'SMS']

X = df['SMS']
y = df['Labels']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=50)

cv = CountVectorizer(stop_words='english')
X_train_cv = cv.fit_transform(X_train)
X_test_cv = cv.transform(X_test)

nb_cv = MultinomialNB()
nb_cv.fit(X_train_cv, y_train)
y_pred_cv = nb_cv.predict(X_test_cv)

print("--- CountVectorizer Results ---")
print(f"Accuracy: {accuracy_score(y_test, y_pred_cv):.4f}")
print("\nClassification Report:\n", classification_report(y_test, y_pred_cv))

--- CountVectorizer Results ---
Accuracy: 0.9830

Classification Report:
               precision    recall  f1-score   support

         ham       0.98      1.00      0.99       954
        spam       0.98      0.90      0.94       161

    accuracy                           0.98      1115
   macro avg       0.98      0.95      0.96      1115
weighted avg       0.98      0.98      0.98      1115



## Part 2: Multinomial Naive Bayes with TF-IDF Vectorizer

In [2]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Feature Extraction: TfidfVectorizer
tfidf = TfidfVectorizer(stop_words='english')
X_train_tfidf = tfidf.fit_transform(X_train)
X_test_tfidf = tfidf.transform(X_test)

# Training
nb_tfidf = MultinomialNB()
nb_tfidf.fit(X_train_tfidf, y_train)
y_pred_tfidf = nb_tfidf.predict(X_test_tfidf)

print("--- TF-IDF Vectorizer Results ---")
print(f"Accuracy: {accuracy_score(y_test, y_pred_tfidf):.4f}")
print("\nClassification Report:\n", classification_report(y_test, y_pred_tfidf))

--- TF-IDF Vectorizer Results ---
Accuracy: 0.9605

Classification Report:
               precision    recall  f1-score   support

         ham       0.96      1.00      0.98       954
        spam       1.00      0.73      0.84       161

    accuracy                           0.96      1115
   macro avg       0.98      0.86      0.91      1115
weighted avg       0.96      0.96      0.96      1115



CountVectorizer proves to be the more optimal feature extraction method. Although TF-IDF provides more spam precision, CountVectorizer resulted in a significantly higher overall accuracy (98.30% vs. 96.05%) and a much better spam recall (0.90 vs. 0.73). In spam detection tasks, catching most spam messages (higher recall) is crucial, making CountVectorizer's raw frequency tracking more effective than TF-IDF's document frequency scaling